src/base/static/components/complex-input-fields/geocoding-field.js

```diff
@@ -0,0 +1,103 @@
+import React, { Component } from "react";
+import PropTypes from "prop-types";
+import classNames from "classnames";
+
+import { geocodingField as messages } from "../messages";
+import "./geocoding-field.scss";
+
+// TODO: Consolidate Util methods used here.
+const Util = require("../../js/utils.js");
+
+class GeocodingField extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      isGeocoding: false,
+      isWithGeocodingError: false,
+    };
+    this.geocodingEngine = this.props.mapConfig.geocoding_engine || "MapQuest";
+    this.hint =
+      this.props.mapConfig.geocode_bounding_box ||
+      this.props.mapConfig.geocode_hint;
+  }
+
+  componentDidMount() {
+    const target = document.getElementsByClassName(
+      "geocoding-field__geocoding-spinner"
+    )[0];
+    // TODO: Replace spin.js spinner with a React component.
+    new Spinner(Shareabouts.smallSpinnerOptions).spin(target);
+  }
+
+  doGeocode() {
+    this.setState({
+      isGeocoding: true,
+      isWithGeocodingError: false,
+    });
+    let address = this.props.value;
+
+    Util[this.geocodingEngine].geocode(address, this.hint, {
+      success: data => {
+        let locationsData = data.results[0].locations;
+        if (locationsData.length > 0) {
+          this.setState({
+            isGeocoding: false,
+            isWithGeocodingError: false,
+          });
+          this.props.emitter.emit("geocode", locationsData[0]);
```

Comment:
I think it needs to be instantiated in the AppView since we have to add the listener there for the `geocode` event? Or would it be better to use a callback function passed from the AppView?

Reply:
Have you tried simply importing and instantiating it here, as well as the app view? Then you don't have to pass it around.

Reply:
I might be missing something about how the emitter works, but don't we need a single instance of it to successfully bus events? Instantiating separate versions of it didn't work for me.

Reply:
We can probably wrap it in a singleton, and import it directly into the components that need it. That'll save the hassle of passing it around.

Happy to explain more on the singleton pattern tonight.

Reply:
@goldpbear You can try this for the singleton pattern:

```
import { EventEmitter } from "fbemitter"

let emitter = new EventEmitter()
console.log("instantiated EventEmitter!")

export default emitter
```

Try importing that into your modules instead of passing around the emitter. You should see that console.log print only once.